### PR TITLE
Green tea m us2

### PR DESCRIPTION
**Pull request by PiotrKowandy**

### CODE REVIEW

File: src/CourseDashboard.cpp [hunk 1]
```diff
@@ -82,3 +83,55 @@ void CourseDashboard::updateUser(User & user)
...
     }
 }
+
+void CourseDashboard::saveToFile(User & user)
+{
+    std::fstream file;
+    try {
+    file.open("../src/UsersList.txt",std::ios::app);
+    file << user.getAllInfoToFile() << std::endl;
+    file.close();
```
Comment (ziobron) [inline src/CourseDashboard.cpp]: formatting - indentations
You don't beed to call `close()`. fstream implements RAII, it closes automatically in destructor.

File: src/CourseDashboard.cpp [hunk 2]
```diff
@@ -82,3 +83,55 @@ void CourseDashboard::updateUser(User & user)
         break;
     }
 }
+
+void CourseDashboard::saveToFile(User & user)
+{
+    std::fstream file;
+    try {
+    file.open("../src/UsersList.txt",std::ios::app);
```
Comment (ziobron) [inline src/CourseDashboard.cpp]: maybe a filename should be a param?

File: src/CourseDashboard.cpp [hunk 3]
```diff
@@ -82,3 +83,55 @@ void CourseDashboard::updateUser(User & user)
...
+    file.close();
+    }
+    catch (std::ifstream::failure e) {
+        std::cerr << "Error while opening file: " << e.what() << std::endl;
+    }
+}
+
+void CourseDashboard::loadFromFile()
+{
+    std::ifstream file("../src/UsersList.txt");
```
Comment (ziobron) [inline src/CourseDashboard.cpp]: filename as a param?

File: src/CourseDashboard.hpp [hunk 4]
```diff
@@ -12,6 +12,8 @@ class CourseDashboard
     void deleteUserByNick(std::string nick);
     void retriveUserByNick(std::string nick);
     void updateUser(User & user);
+    void saveToFile(User & user);
+    void loadFromFile();
```
Comment (ziobron) [inline src/CourseDashboard.hpp:16]: loadFromFile loads all users, but saveToFile saves only one. How to save all users?

File: src/User.hpp [hunk 5]
```diff
@@ -6,13 +6,15 @@
 class User
 {
 public:
+    User() = default;
```
Comment (ziobron) [inline src/User.hpp]: We don't want this constructor. What is a default user?

File: src/main.cpp [hunk 6]
```diff
@@ -6,12 +6,14 @@ int main()
...
-    std::cout<<"after delete: "<<std::endl;
+    cdb.loadFromFile();
+    //User u1("Szymon", "SzymonGajewski", "wieczorowa", "SzymonGajewski", "SzymonGajewski");
+    //cdb.createUser(u1);
+    //cdb.retriveUserByNick("SzymonGajewski");
     cdb.showAll();
+    //cdb.saveToFile(u1);
+    //cdb.deleteUserByNick("SzymonGajewski");
+    //std::cout<<"after delete: "<<std::endl;
+    //cdb.showAll();
```
Comment (ziobron) [inline src/main.cpp]: no comments in a production code.

File: tests/CourseDashboardTests.cpp [hunk 7]
```diff
@@ -1,8 +1,25 @@
...
+    CourseDashboard testCdb;
+    User testU1;
+
+    CourseDashboardTests()
+    {
+        testU1 = User("Szymon", "SzymonGajewski", "wieczorowa", "SzymonGajewski", "SzymonGajewski");
+        testCdb.createUser(testU1);
+        testCdb.retriveUserByNick("SzymonGajewski");
+        testCdb.deleteUserByNick("SzymonGajewski");
+        EXPECT_EQ("SzymonGajewski", testU1.getNick());
```
Comment (ziobron) [inline tests/CourseDashboardTests.cpp]: this EXPECT_EQ doesn't test anything in the CourseDashboard. It only checks if the User nick was set properly. This is not a test of CourseDashboard.

**Review by ziobron — CHANGES_REQUESTED** [review]
No tests :(

File: src/CourseDashboard.cpp (hunk 1 above)
Comment (PiotrKowandy) [reply src/CourseDashboard.cpp]: Corrected - close function removed.

File: src/CourseDashboard.cpp (hunk 2 above)
Comment (PiotrKowandy) [reply src/CourseDashboard.cpp]: Param added, formatting fixed.

File: src/CourseDashboard.hpp (hunk 4 above)
Comment (PiotrKowandy) [reply src/CourseDashboard.hpp:16]: Currently all users are provided into file.

File: src/User.hpp (hunk 5 above)
Comment (PiotrKowandy) [reply src/User.hpp]: User() with default it was stupid idea. Constructor removed.

File: src/main.cpp (hunk 6 above)
Comment (PiotrKowandy) [reply src/main.cpp]: No comments in files.

File: tests/CourseDashboardTests.cpp (hunk 7 above)
Comment (PiotrKowandy) [reply tests/CourseDashboardTests.cpp]: Yes it's true, These EXPECTs weren't tests.

Comment (PiotrKowandy) [reply tests/CourseDashboardTests.cpp]: Changed on ASSERT_TRUE.

File: src/CourseDashboard.cpp [hunk 8]
```diff
@@ -13,12 +14,21 @@ void CourseDashboard::showAll()
...
-void CourseDashboard::createUser(const User & user)
+bool CourseDashboard::createUser(const User & user)
 {
-    users_.emplace_back(user);
+    try
+    {   
+        users_.emplace_back(user);
+        return true;
+    }
+    catch (...)
```
Comment (ziobron) [inline src/CourseDashboard.cpp:24]: catching everything is not advised. It may be skipped here.

File: src/CourseDashboard.cpp [hunk 9]
```diff
@@ -13,12 +14,21 @@ void CourseDashboard::showAll()
...
+    {
+        return false;
+    }
+        
 }
 
-void CourseDashboard::deleteUserByNick(std::string nick)
+bool CourseDashboard::deleteUserByNick(std::string nick)
 {  
     auto it = std::find_if(std::begin(users_), std::end(users_), [nick](const auto & user)
```
Comment (ziobron) [inline src/CourseDashboard.cpp:33]: maybe std::remove_if?

File: src/CourseDashboard.cpp [hunk 10]
```diff
@@ -82,3 +95,66 @@ void CourseDashboard::updateUser(User & user)
         break;
     }
 }
+
+void CourseDashboard::saveUsersToFile(std::vector<User>  users)
```
Comment (ziobron) [inline src/CourseDashboard.cpp:99]: why do we have a parameter here? users vector is available inside CourseDashboard class.

File: src/CourseDashboard.cpp [hunk 11]
```diff
@@ -82,3 +95,66 @@ void CourseDashboard::updateUser(User & user)
...
+{
+    std::fstream file;
+    std::string fileName = "../src/UsersList.txt";
```
Comment (ziobron) [inline src/CourseDashboard.cpp:102]: fileName could a parameter to this function

File: src/CourseDashboard.cpp [hunk 12]
```diff
@@ -82,3 +95,66 @@ void CourseDashboard::updateUser(User & user)
...
+    }
+    catch (std::ifstream::failure e) 
+    {
+        std::cerr << "Error while opening file: " << e.what() << std::endl;
+    }
+}
+
+void CourseDashboard::loadFromFile()
+{
+    std::string fileName = "../src/UsersList.txt";
```
Comment (ziobron) [inline src/CourseDashboard.cpp:119]: fileName could a parameter to this function

File: src/User.cpp [hunk 13]
```diff
@@ -25,6 +25,17 @@ std::string User::getAllInfo() const
...
+{
+    std::stringstream ss;
+    ss<<std::left;
+    ss<<std::setw(1) << name_ << std::endl;
+    ss<<std::setw(1) << nick_ << std::endl;
+    ss<<std::setw(1) << group_ << std::endl;
+    ss<<std::setw(1) << gitHub_ << std::endl;
+    ss<<std::setw(1) << firecode_ << std::endl;
+    return ss.str();
+}
```
Comment (ziobron) [inline src/User.cpp:38]: hmm... isn't this almost the same as getAllInfo()? Maybe one of those functions is redundant? It's code duplication.

**Review by ziobron — CHANGES_REQUESTED** [review]
Still missing tests

**Comment by ziobron** [comment]
+7 points (no tests)

**Pull request closed without merging**